packages/steveo/src/runtime/workflow.ts

```diff
@@ -0,0 +1,210 @@
+   * it will load the current state and execute the step, or continue
+   * executing the rollback chain if there was an unrecoverable error.
+   *
+   * (Yes the name is confusing, I'm following precedent 😁)
+   */
+  async subscribe<T extends WorkflowPayload>(payload: T) {
+    return this.execute(payload);
+  }
+
+  /**
+   *
+   */
+  async execute<T extends WorkflowPayload>(payload: T) {
+
+    await this.storage.transaction(async (transaction) => {
+      const workflowId = payload.workflowId ?? `${this._name}-${v4()}`;
+
+      if (!payload.workflowId) {
+        // TODO: Save new workflow state in DB
+        // await this.db.
+      }
+
+      await this.executeForward(workflowId, payload, transaction);
+    });
+  }
+
+  /**
+   *
+   * @param payload
+   */
+  async publish<T>(payload: T | T[], context?: { key: string }) {
+
+    await this.storage.transaction(async () => {
+      const params = Array.isArray(payload) ? payload : [payload];
+
+      try {
+        // sqs calls this method twice
+        await this._producer.initialize(this.topic);
+
+        await Promise.all(
+          params.map((data: T) => {
+            this._registry.emit('task_send', this.topic, data);
+            return this._producer.send(this.topic, data, context?.key, context);
+          })
+        );
+
+        this._registry.emit('task_success', this.topic, payload);
+      }
+      catch (err) {
+        this._registry.emit('task_failure', this.topic, err);
+        throw err;
+      }
+    });
+
+  }
+
+  /**
+   *
+   */
+  private async executeForward(workflowId: string, payload: unknown, transaction: TransactionHandle): Promise<void> {
+
+    if (!this.steps?.length) {
+      throw new Error(`Steps must be defined before a flow is executed ${this._name}`);
+    }
+
+    const flowState = await this.loadState(workflowId);
+
+    try {
+
+      if (flowState.current < 0) {
+        throw new Error(`Workflow ${workflowId} step ${flowState.current} cannot be less than zero`);
+      }
+
+      if (flowState.current > flowState.results.length) {
+        throw new Error(`Workflow ${workflowId} step ${flowState.current} exceeds available steps ${flowState.results.length}`)
+      }
+
+      // If this state has failed then we pass control over to the
+      // rollback executor.
+      if (flowState.failedStep) {
+        return this.rollbackContinue(flowState, transaction);
+      }
+
+      // TODO: Add `source` to the workflow state to check the current steveo instance service name it is running in  and prevent accidental network boundary crossing
+
+      // TODO: Protect out of order excution or step re-execution
+
+      const step = this.steps[flowState.current];
+
+      // TODO: Further step validation (What? what were you thinking Paul? 🤦)
+
+      const result = await step.execute(payload);
+
+      // TODO: Update state
+
+      flowState.results[flowState.current] = result;
+      flowState.current++;
```

Comment:
Using integers to decide what step we're up to is going to introduce some funkiness if anyone edits the steps in the workflow. Any reason we can't use the step name? The name could default to the task name, and we can allow callers to provide a different name in the cases where they want to have repeated steps. 

would also make debugging easier as you can read the name of the step and the result

Reply:
Sounds like a good idea

Reply:
https://github.com/ordermentum/steveo/pull/800/commits/1e1234041b1cf44adb3516b7f5c760972f5da4f7